### PR TITLE
Rewrite part of file with StdlibUnittest

### PR DESCRIPTION
**Pull request by frootloops**

### CODE REVIEW

**Comment by gottesmm** [comment]
Is it possible for you to also split up this file into tests for different sorts of printing (i.e. one for Array, one for Integer, one for Floating Point)?

Very long stdlib unittest tests are very difficult to debug when we run into optimizer issues and splitting it into smaller files would make it easier to bisect/debug.

**Comment by frootloops** [comment]
@gottesmm  Yes, It's possible. Do you want to see it in one big PR or in a few small PRs?

**Comment by gottesmm** [comment]
I would suggest a few small PRs. It will make it easier to localize/fix/revert commits if issues come up.

**Comment by frootloops** [comment]
@gottesmm  Ok. So is this PR ready to merge? :)

**Comment by gottesmm** [comment]
I actually think I need someone on the stdlib team to review it. (I was going to review it myself, but I am not a stdlib engineer). I am going to send it to Dmitri.

**Comment by frootloops** [comment]
@gottesmm thanks. I will continue to work on this file

**Comment by frootloops** [comment]
Should I create `test/1_stdlib/Print` folder or `test/1_stdlib/PrintArray.swif` etc when I will split this big file into small files?

**Comment by gottesmm** [comment]
I am not sure TBH. I would leave that up to @gribozavr.

That being said, I imagine that if you put the tests in the Print folder, it will be ok to call the tests Print*.swift. So why not just do it in test/1_stdlib and if Dmitri wants us to create a folder, then we can move the tests to that directory in a subsequent commit. That would make it easier in terms of munging patches.

**Comment by gribozavr** [comment]
@frootloops Thanks!  Two comments:

1. You need to call `runAllTests()`, or none of the new tests would be run.

2. Because StdlibUnittest forks a new process, I don't think the command line argument would be passed down to it.  I think the way to do it would be to change the `_ParentProcess._spawnChild()` method to also forward other command line arguments.  You'd need to add a test for that in a new file, `validation-test/stdlib/StdlibUnittestCommandLineArguments.swift`.

**Comment by frootloops** [comment]
@gribozavr I'm still working on PR but if you can please have a look and leave any comments.

**Comment by frootloops** [comment]
@gribozavr when you will have time please have a look :)

**Comment by frootloops** [comment]
```
Testing Time: 1415.60s
  Expected Passes    : 7668
  Expected Failures  : 8
  Unsupported Tests  : 39
-- check-swift-all-macosx-x86_64 finished --
--- Finished tests for swift ---
```

**Comment by gribozavr** [comment]
@frootloops Thank you!

> 1. I have a lot of copies `StructPrintable`, `LargeStructPrintable` and others. How can I share these structs and classes between test files or maybe it's ok.

Swift can share code with modules.  You can put common definitions into, say, test/1_stdlib/Inputs/PrintTestTypes.swift, and build modules out of it in every test that needs these types.

> 2. Should I put `// XFAIL: linux` in every new `Print*` files?

Only if it fails on Linux.  Let me know if I need to run these for you.

> 3. Should I put `setlocale(LC_ALL, arg)` stuff in every new `Print*` files?

Strictly speaking, only floating point tests need it.  We know from the implementation that other behavior is implemented without relying on locales.

> 4. What is the difference  between `RUN: %target-run` and `RUN: %target-run-simple-swift`?

`docs/Testing.rst`:

> * `%target-run`: run a command on the target machine.
>
>* `%target-run-simple-swift`: build a one-file Swift program and run it on the target machine.

So, the latter builds the binary for you, while for the former you need to build it yourself using another `RUN` line.

File: test/1_stdlib/PrintArray.swift [hunk 1]
```diff
@@ -0,0 +1,98 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import Swift
+import Darwin
```
Comment (gribozavr) [inline test/1_stdlib/PrintArray.swift]: I don't think anything depends on `Darwin` here.

File: test/1_stdlib/PrintBoolean.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import Swift
+import Darwin
```
Comment (gribozavr) [inline test/1_stdlib/PrintBoolean.swift]: I don't think anything depends on `Darwin` here.

File: test/1_stdlib/PrintClass.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import Swift
+import Darwin
```
Comment (gribozavr) [inline test/1_stdlib/PrintClass.swift]: I don't think anything depends on `Darwin` here.

File: test/1_stdlib/PrintDictionary.swift [hunk 4]
```diff
@@ -0,0 +1,36 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import Swift
+import Darwin
```
Comment (gribozavr) [inline test/1_stdlib/PrintDictionary.swift]: I don't think anything depends on `Darwin` here.

**Comment by gribozavr** [comment]
@frootloops Please try to prune `import Darwin` from the smaller files, most of them don't need it as far as I see.  This will allow the tests to pass on Linux.

**Comment by frootloops** [comment]
@gribozavr have a look at the PR. I updated it and if you can please run a test suite on a linux machine.

File: test/1_stdlib/Inputs/PrintTestTypes.swift [hunk 5]
```diff
@@ -0,0 +1,153 @@
+import Swift
+import Foundation
```
Comment (gribozavr) [inline test/1_stdlib/Inputs/PrintTestTypes.swift]: It is not required to import Swift explicitly.  Why do we import Foundation?  We can't do it on Linux in the swift repository.

File: test/1_stdlib/Inputs/PrintTestTypes.swift [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    self.x = x
+  }
+
+  public var description: String {
+    return "►\(x)◀︎"
+  }
+}
+
+public struct LargeStructPrintable : CustomStringConvertible,
+ProtocolUnrelatedToPrinting {
```
Comment (gribozavr) [inline test/1_stdlib/Inputs/PrintTestTypes.swift]: Please add two spaces of indentation for continuation lines:

```swift
public struct LargeStructPrintable :
  CustomStringConvertible,
  ProtocolUnrelatedToPrinting {
```

File: test/1_stdlib/Inputs/PrintTestTypes.swift [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        self.x = x
+  }
+
+  public var debugDescription: String {
+        return "►\(x)◀︎"
+  }
+}
+
+public struct StructVeryPrintable : CustomStringConvertible,
+CustomDebugStringConvertible, ProtocolUnrelatedToPrinting {
```
Comment (gribozavr) [inline test/1_stdlib/Inputs/PrintTestTypes.swift]: And here.

**Comment by frootloops** [comment]
@gribozavr ready

**Comment by frootloops** [comment]
@gribozavr BTW: I saw - https://twitter.com/nadavrot/status/675495662989893633 and if you know something about the internship. What are the requirements for candidates?

**Comment by gribozavr** [comment]
@frootloops As far as I know, you need to be enrolled in full-time school the next year after the internship.  There might be more formal requirements that I don't know about, please ask swift-intern@group.apple.com for details.

**Comment by frootloops** [comment]
@gribozavr Ok. Thank you!

File: test/1_stdlib/Print.swift [hunk 8]
```diff
@@ -1,1045 +1,75 @@
-// RUN: mkdir -p %t
-// RUN: %target-build-swift %s -parse-stdlib -Xfrontend -disable-access-control -o %t/a.out -Xlinker -dead_strip
-// RUN: %target-run %t/a.out env | FileCheck %s
-// RUN: %target-run %t/a.out ru_RU.UTF-8 | FileCheck %s
+// RUN: rm -rf %t && mkdir %t
+// RUN: %target-build-swift -emit-library -Xfrontend -enable-resilience -c  %S/Inputs/PrintTestTypes.swift -o %t/PrintTestTypes.o
+// RUN: %target-build-swift -emit-module -Xfrontend -enable-resilience -c  %S/Inputs/PrintTestTypes.swift -o %t/PrintTestTypes.o
```
Comment (gribozavr) [inline test/1_stdlib/Print.swift]: This is wasteful, you can build both with one command:

```
-c -force-single-frontend-invocation -parse-as-library -emit-module -emit-module-path ./foo.swiftmodule -o foo.o foo.swift
```

File: test/1_stdlib/PrintArray.swift [hunk 9]
```diff
@@ -0,0 +1,38 @@
+// RUN: rm -rf %t && mkdir %t
+// RUN: %target-build-swift -emit-library -Xfrontend -enable-resilience -c  %S/Inputs/PrintTestTypes.swift -o %t/PrintTestTypes.o
+// RUN: %target-build-swift -emit-module -Xfrontend -enable-resilience -c  %S/Inputs/PrintTestTypes.swift -o %t/PrintTestTypes.o
```
Comment (gribozavr) [inline test/1_stdlib/PrintArray.swift]: Same here (and everywhere).

File: test/1_stdlib/PrintPointer.swift [hunk 10]
```diff
@@ -0,0 +1,38 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import Swift
```
Comment (gribozavr) [inline test/1_stdlib/PrintPointer.swift]: No need to `import Swift`.

File: test/1_stdlib/Print.swift [hunk 11]
```diff
@@ -1,1045 +1,75 @@
...
+// RUN: %target-build-swift %s -Xlinker %t/PrintTestTypes.o -I %t -L %t -o %t/main
+// RUN: %target-run %t/main
 // REQUIRES: executable_test
 
-// XFAIL: linux
-
 import Swift
```
Comment (gribozavr) [inline test/1_stdlib/Print.swift]: No need to `import Swift`.

File: test/1_stdlib/PrintArray.swift [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+// RUN: %target-build-swift %s -Xlinker %t/PrintTestTypes.o -I %t -L %t -o %t/main
+// RUN: %target-run %t/main
+// REQUIRES: executable_test
+
+import Swift
```
Comment (gribozavr) [inline test/1_stdlib/PrintArray.swift]: No need to `import Swift`.

File: test/1_stdlib/PrintFloat.swift [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+import Darwin
+import PrintTestTypes
+
+let PrintTests = TestSuite("PrintFloat")
+
+let arg = Process.arguments[1]
+if arg == "env" {
+  setlocale(LC_ALL, "")
+} else {
+  expectEqual("ru_RU.UTF-8", arg)
```
Comment (gribozavr) [inline test/1_stdlib/PrintFloat.swift]: I'm sorry, but are you sure this works with StdlibUnittest?  I don't think it passes down the arguments to the test runner process.  You would need to modify StdlibUnittest to do so, and here you would search for the argument anywhere among the command line parameters.

**Comment by gribozavr** [comment]
Looks good in general, but I'm concerned about the locale thing, could you verify?

File: test/1_stdlib/PrintPointer.swift [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  expectPrinted(expectedNull, nullUP)
+  
+  expectPrinted("UnsafeBufferPointer(start: \(expectedNull), length: 0)",
+    UnsafeBufferPointer(start: nullUP, count: 0))
+  expectPrinted("UnsafeMutableBufferPointer(start: \(expectedNull), length: 0)",
+    UnsafeMutableBufferPointer(start: nullUP, count: 0))
+  
+  expectPrinted(expectedNull, COpaquePointer())
+  expectPrinted(expectedNull, CVaListPointer(_fromUnsafeMutablePointer: nullUP))
+  expectPrinted(expectedNull, AutoreleasingUnsafeMutablePointer<Int>())
```
Comment (gribozavr) [inline test/1_stdlib/PrintPointer.swift]: This line fails on Linux:

```
error: use of unresolved identifier 'AutoreleasingUnsafeMutablePointer'
```

Please guard it with `#if _runtime(_ObjC)`.

**Comment by frootloops** [comment]
@gribozavr  You were right about locale. Can you run tests on linux again?

File: test/1_stdlib/PrintFloat.swift [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+// REQUIRES: executable_test
+// XFAIL: linux
+
+import StdlibUnittest
+import Darwin
+import PrintTestTypes
+
+let PrintTests = TestSuite("PrintFloat")
+
+PrintTests.test("Locale") {
```
Comment (gribozavr) [inline test/1_stdlib/PrintFloat.swift]: This should be modeled as a setup block.

File: stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb [hunk 16]
```diff
@@ -803,7 +814,10 @@ public func runAllTests() {
 "                        Useful for running under a debugger.\n" +
 "--stdlib-unittest-filter FILTER-STRING\n" +
 "                        only run tests whose names contain FILTER-STRING as\n" +
-"                        a substring."
+"                        a substring.\n" + 
+"--env STRING\n" +
```
Comment (gribozavr) [inline stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb]: Could you just pass through all arguments (that don't start with `--stdlib-unittest`) to the child process?  This is how it was intended to be implemented.

**Comment by frootloops** [comment]
@gribozavr thanks for comments. I updated the PR

**Comment by gribozavr** [comment]
@frootloops Thank you, merged!

**Pull request merged**